pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -0,0 +1,1455 @@
+            assert(index >= 0);
+            size_t rankId =
+                static_cast<size_t>(index) >> BaseType::getNumQubits();
+
+            if (rankId == mpi_manager_.getRank()) {
+                int local_index =
+                    static_cast<size_t>(
+                        rankId * std::pow(2.0, static_cast<long double>(
+                                                   BaseType::getNumQubits()))) ^
+                    index;
+                indices_local.push_back(local_index);
+                values_local.push_back(values[i]);
+            }
+        }
+
+        auto device_id = BaseType::getDataBuffer().getDevTag().getDeviceID();
+        auto stream_id = BaseType::getDataBuffer().getDevTag().getStreamID();
+
+        index_type num_elements = indices_local.size();
+
+        DataBuffer<index_type, int> d_indices{
+            static_cast<std::size_t>(num_elements), device_id, stream_id, true};
+
+        DataBuffer<CFP_t, int> d_values{static_cast<std::size_t>(num_elements),
+                                        device_id, stream_id, true};
+
+        d_indices.CopyHostDataToGpu(indices_local.data(), d_indices.getLength(),
+                                    async);
+        d_values.CopyHostDataToGpu(values_local.data(), d_values.getLength(),
+                                   async);
+
+        setStateVector_CUDA(BaseType::getData(), num_elements,
+                            d_values.getData(), d_indices.getData(),
+                            thread_per_block, stream_id);
+        PL_CUDA_IS_SUCCESS(cudaDeviceSynchronize());
+        mpi_manager_.Barrier();
+    }
+
+    /**
+     * @brief Apply a single gate to the state-vector. Offloads to custatevec
+     * specific API calls if available. If unable, attempts to use prior cached
+     * gate values on the device. Lastly, accepts a host-provided matrix if
+     * otherwise, and caches on the device for later reuse.
+     *
+     * @param opName Name of gate to apply.
+     * @param wires Wires to apply gate to.
+     * @param adjoint Indicates whether to use adjoint of gate.
+     * @param params Optional parameter list for parametric gates.
+     */
+    void applyOperation(
+        const std::string &opName, const std::vector<size_t> &wires,
+        bool adjoint = false, const std::vector<Precision> &params = {0.0},
+        [[maybe_unused]] const std::vector<CFP_t> &gate_matrix = {}) {
+        const auto ctrl_offset = (BaseType::getCtrlMap().find(opName) !=
+                                  BaseType::getCtrlMap().end())
+                                     ? BaseType::getCtrlMap().at(opName)
+                                     : 0;
+        const std::vector<std::size_t> ctrls{wires.begin(),
+                                             wires.begin() + ctrl_offset};
+        const std::vector<std::size_t> tgts{wires.begin() + ctrl_offset,
+                                            wires.end()};
+        if (opName == "Identity") {
```

Comment:
Out of curiosity, in cases where gate operations are really fast, isn't there any penalty using string comparisons instead of enums in such switch sequences?

Reply:
Maybe it's never really that fast, but I've once seen a bottleneck show up in sparse matrix routines that would branch based on "coo"/"csc" formats, which was solved by switching to integers ;)

Reply:
I wouldn't worry too much about it here since the time spent on string comparison will be almost nothing compared to the time spent on gate operations, maybe it will be even faster than the overhead of GPU kernel launchment.